# Задание 3
## Вы работаете продуктовым аналитиком в агрегаторе такси SkyTaxi, и ваша задача - оценить результаты АВ-теста.
### Ваша задача - исследовать результат и понять, влияет ли понижение цены на конверсию из просмотра в создание заказа.

In [2]:
import pandas as pd
import numpy as np
import datetime as dt
from scipy.stats import ttest_ind

## Задание 3.1
Импортируйте файл с данными в Jupyter Notebook и расшифруйте идентификаторы городов с помощью листа city_dict

In [4]:
dts = pd.read_excel('Финальная первичная аттестация. Вариант 1. Python.xlsx',sheet_name= 'Dataset')
city = pd.read_excel('Финальная первичная аттестация. Вариант 1. Python.xlsx',sheet_name= 'city_dict')

In [5]:
new_dts = dts.merge(city, on='id_city',how= 'inner')
new_dts

,id_view,time_view,id_group,id_city,nflag_order,name_city
0,10457162393,2023-01-17 13:56:06.118,0,3,1,Новосибирск
1,10457197478,2023-01-02 11:02:58.189,0,3,0,Новосибирск
2,10457135372,2023-01-15 10:03:58.011,0,3,0,Новосибирск
3,10457162971,2023-01-02 09:39:31.283,0,3,1,Новосибирск
4,10457132942,2023-01-23 19:18:03.516,0,3,0,Новосибирск
...,...,...,...,...,...,...
21934,10457174004,2023-01-21 11:47:30.578,0,9,1,Воронеж
21935,10457130320,2023-01-24 02:09:49.169,0,9,0,Воронеж
21936,10457166866,2023-01-13 16:22:38.491,1,9,0,Воронеж
21937,10457186695,2023-01-28 02:21:49.007,0,9,1,Воронеж


## Задание 3.2
Рассчитайте результаты АВ-теста и с помощью статистического критерия сделайте вывод

In [7]:
s, p = ttest_ind(new_dts[new_dts['id_group'] == 0]["nflag_order"], new_dts[new_dts['id_group'] == 1]["nflag_order"])
print ("stat:",s)
print ("p_value:",p)

stat: -9.41595134321716
p_value: 5.1365957325220535e-21


Гипотеза H0 не подтвердилась, так как средние не равны.


## Задание 3.3
Запустите цикл, который пробежится по всем городам в датасете. Выявите те города, в которых есть статистически значимая разница между группами А и В. Создайте переменную типа "список", которая будет хранить эти города.

In [10]:
ttest_new_list = []
for i in new_dts['name_city'].unique():
  s, p = ttest_ind(new_dts[(new_dts['name_city']==i) & (new_dts['id_group']==0)]['nflag_order'], new_dts[(new_dts['name_city']==i) & (new_dts['id_group']==1)]['nflag_order'])
  if p < 0.05:
    ttest_new_list.append(i)

In [11]:
ttest_new_list

['Москва', 'Санкт-Петербург', 'Казань']

## Задание 3.4
Создайте дополнительную колонку со значением 0 или 1 в зависимости от часа заказа.

1 - если час пик (возьмите, например, 7,8,9,10,17,18,19,20), а 0 - все остальные часы.

Проверьте результат АВ-теста по отдельности на пиковых и не-пиковых заказах и сделайте выводы.

In [13]:
new_dts['hour'] = new_dts['time_view'].dt.hour
new_dts

,id_view,time_view,id_group,id_city,nflag_order,name_city,hour
0,10457162393,2023-01-17 13:56:06.118,0,3,1,Новосибирск,13
1,10457197478,2023-01-02 11:02:58.189,0,3,0,Новосибирск,11
2,10457135372,2023-01-15 10:03:58.011,0,3,0,Новосибирск,10
3,10457162971,2023-01-02 09:39:31.283,0,3,1,Новосибирск,9
4,10457132942,2023-01-23 19:18:03.516,0,3,0,Новосибирск,19
...,...,...,...,...,...,...,...
21934,10457174004,2023-01-21 11:47:30.578,0,9,1,Воронеж,11
21935,10457130320,2023-01-24 02:09:49.169,0,9,0,Воронеж,2
21936,10457166866,2023-01-13 16:22:38.491,1,9,0,Воронеж,16
21937,10457186695,2023-01-28 02:21:49.007,0,9,1,Воронеж,2


In [14]:
time = [7,8,9,10,17,18,19,20]
new_dts['flag_hour_peak'] = np.where(new_dts['hour'].isin(time) ,1,0)
new_dts

,id_view,time_view,id_group,id_city,nflag_order,name_city,hour,flag_hour_peak
0,10457162393,2023-01-17 13:56:06.118,0,3,1,Новосибирск,13,0
1,10457197478,2023-01-02 11:02:58.189,0,3,0,Новосибирск,11,0
2,10457135372,2023-01-15 10:03:58.011,0,3,0,Новосибирск,10,1
3,10457162971,2023-01-02 09:39:31.283,0,3,1,Новосибирск,9,1
4,10457132942,2023-01-23 19:18:03.516,0,3,0,Новосибирск,19,1
...,...,...,...,...,...,...,...,...
21934,10457174004,2023-01-21 11:47:30.578,0,9,1,Воронеж,11,0
21935,10457130320,2023-01-24 02:09:49.169,0,9,0,Воронеж,2,0
21936,10457166866,2023-01-13 16:22:38.491,1,9,0,Воронеж,16,0
21937,10457186695,2023-01-28 02:21:49.007,0,9,1,Воронеж,2,0


In [15]:
for cy in new_dts['name_city'].unique():
  print(cy)
  print()
  #Пиковые
  s, p = ttest_ind(new_dts[(new_dts['name_city']==cy) & (new_dts['flag_hour_peak'] == 1) & (new_dts['id_group']==0)]['nflag_order'],
                     new_dts[(new_dts['name_city']==cy) & (new_dts['flag_hour_peak'] == 1) & (new_dts['id_group']==1)]['nflag_order'])
  print ("Пиковые заказы:")
  print ("stat:", s)
  print ("p_value:", p)
  if p < 0.05:
    print("Гипотеза Н0 не верна. Между группами есть разница. Ошибка выборки минимальна.")
  else:
      print("Гипотеза Н0 побеждает. Между группами нет разницы или мало данных.")
  print()
  
  #Не-пиковые
  s_t, p_t = ttest_ind(new_dts[(new_dts['name_city']==cy) & (new_dts['flag_hour_peak'] == 0) & (new_dts['id_group']==0)]['nflag_order'],
                     new_dts[(new_dts['name_city']==cy) & (new_dts['flag_hour_peak'] == 0) & (new_dts['id_group']==1)]['nflag_order'])
  print ("Не пиковые заказы:")
  print ("stat:", s_t)
  print ("p_value:", p_t)
  if p_t < 0.05:
    print("Гипотеза Н0 не верна. Между группами есть разница. Ошибка выборки минимальна.")
  else:
      print("Гипотеза Н0 побеждает. Между группами нет разницы или мало данных.")
  print()
  print()
  print()
 

Новосибирск

Пиковые заказы:
stat: 0.03509071737172927
p_value: 0.9720136785031339
Гипотеза Н0 побеждает. Между группами нет разницы или мало данных.

Не пиковые заказы:
stat: -1.9033169219042654
p_value: 0.05713158719650457
Гипотеза Н0 побеждает. Между группами нет разницы или мало данных.



Москва

Пиковые заказы:
stat: -6.790685972172624
p_value: 1.3449585112087216e-11
Гипотеза Н0 не верна. Между группами есть разница. Ошибка выборки минимальна.

Не пиковые заказы:
stat: -4.291499026558841
p_value: 1.8030823524055564e-05
Гипотеза Н0 не верна. Между группами есть разница. Ошибка выборки минимальна.



Санкт-Петербург

Пиковые заказы:
stat: -4.245735352177566
p_value: 2.3018931356634527e-05
Гипотеза Н0 не верна. Между группами есть разница. Ошибка выборки минимальна.

Не пиковые заказы:
stat: -3.407489376223579
p_value: 0.0006634181415393023
Гипотеза Н0 не верна. Между группами есть разница. Ошибка выборки минимальна.



Казань

Пиковые заказы:
stat: -1.2827249444278546
p_value: 0.20

## Задание 3.5
Опишите ваши результаты. Ваши выводы можно оформить внутри тетрадки .ipynb

Сначала мной были импортированы файлы с данными в Jupyter Notebook. Далее я расшифровал идентификаторы городов с помощью листа city_dict.
А после, рассчитав результаты АВ-теста и с помощью статистического критерия, я пришёл к выводу, что Гипотеза H0 не подтвердилась, так как средние не равны.

Запустив цикл, который пробежится по всем городам в датасете, Я выявил те города, в которых есть статистически значимая разница между группами А и В. Этими городами является: "Москва", "Санкт-Петербург", "Казань".

Создав дополнительную колонку со значением 0 или 1 в зависимости от часа заказа.
1 - если час пик (возьмите, например, 7,8,9,10,17,18,19,20), а 0 - все остальные часы.
Далее я проверил результаты АВ-теста по отдельности на пиковых и не-пиковых заказах.